oauthproxy.go

```diff
@@ -1211,6 +1212,28 @@ func checkAllowedGroups(req *http.Request, s *sessionsapi.SessionState) bool {
 	return false
 }
 
+// checkAllowedRoles allow secondary role restrictions based on the `allowed_roles`
+// querystring parameter. This currently only works for Keycloak OIDC.
+func checkAllowedRoles(req *http.Request, s *sessionsapi.SessionState) bool {
+	allowedRoles := extractAllowedEntities(req, "allowed_roles")
+	if len(allowedRoles) == 0 {
+		return true
+	}
+
+	allowedGroups := map[string]struct{}{}
+	for role := range allowedRoles {
+		allowedGroups["role:"+role] = struct{}{}
+	}
```

Comment:
As this is a Keycloak specific feature. I don't like to see the implementation for it in such a generic place of the codebase!

Reply:
To properly do this check. You should override the `Authorize()` method of the `Provider` interface in the Keycloak OIDC provider. Call the base `Authorize` of the `OIDC` provider and add the `allowedRoles` logic on top of it. It should most definitely not be done through the generic OIDC constraints.

Reply:
After a very long while I finally found some time to look into this again, and I am not sure whether what you propose addresses the same issue as I intended to solve.

Generally, there seem to currently be two places where group constraints can be imposed: in the configuration of the provider (https://oauth2-proxy.github.io/oauth2-proxy/configuration/overview?_highlight=allowed&_highlight=group#general-provider-options), limiting who can obtain a session at the proxy, but also on a per-request basis through a query parameter to the `/auth` endpoint (https://oauth2-proxy.github.io/oauth2-proxy/features/endpoints#auth).

An "allowed roles" constraint already exists for Keycloak in the provider configuration, but what I aim to add here is support for the respective query parameter in the `/auth` endpoint - my use case being that I want to deploy a single oauth2 proxy instance for my entire cluster, but have separate applications on it request their own specific role/group constraints.

These per-request constraints do not seem to pass through the provider infrastructure at all - the other checks (email, domain, group) are all re-implemented specifically for this endpoint. Within the `Authorize` function in the provider interfaces, I believe I do not even have access to the request query parameters.

What would be a better way to approach this?